// linking/preact/index.esm.js
import { combineReducers, createStore as createStore$1 } from 'redux';
import { Component, h, render } from 'preact';
import linkFactory, { handleActions, sideEffect } from 'linking';

function composeReduce(reducers) {
  let wrapped;
  wrapped = Object.entries(reducers).map(function ([key, reducer]) {
    return {
      [key]: typeof reducer === "function" ? reducer : handleActions(reducer)
    };
  });
  return combineReducers(Object.assign(...wrapped));
}

function setup({
  composeReduce,
  Component: Component$$1,
  h: h$$1,
  render: render$$1,
  linkFactory: linkFactory$$1,
  sideEffectFactory
}) {
  let link;
  link = linkFactory$$1({
    Component: Component$$1
  });
  return {
    link,
    sideEffect: sideEffectFactory(Component$$1),
    startApp: function ({
      reducers,
      preload,
      component,
      env = window,
      el = "#root"
    }) {
      let state, store, container, root, replaceApp;
      state = (typeof preload == "function" ? preload(env) : void 8) || preload;
      store = createStore$1(composeReduce(reducers), state, env != null ? typeof env.__REDUX_DEVTOOLS_EXTENSION__ == "function" ? env.__REDUX_DEVTOOLS_EXTENSION__() : void 8 : void 8);
      container = env != null ? env.document.querySelector(el) : void 8;
      root = container.firstElementChild;
      return typeof module !== "undefined" && module.hot ? (replaceApp = function (it) {
        root = render$$1(h$$1(link(it), {
          store
        }), container, root);
      }, replaceApp(component), Object.assign({}, store, {
        replaceApp
      }, {
        replaceReducer: function (it) {
          return store.replaceReducer(composeReduce(it));
        }
      })) : (render$$1(h$$1(link(component), {
        store
      }), container, root), store);
    }
  };
}

let sideEffect$1;
let link;
let startApp;
({
  link,
  startApp,
  sideEffect: sideEffect$1
} = setup({
  composeReduce,
  h,
  Component,
  render,
  linkFactory,
  sideEffectFactory: sideEffect
}));

function renderVdom({
  reducers,
  preload,
  component
}) {
  let state, store;
  state = preload({}) || preload;
  store = createStore(composeReduce(reducers), state);
  return h(link(component), {
    store
  });
}

export { startApp as linkStart, startApp, link, h, sideEffect$1 as sideEffect, renderVdom };
export default link;
